vault/logical_system_raft.go

```diff
@@ -236,26 +237,46 @@ func (b *SystemBackend) handleRaftRemovePeerUpdate() framework.OperationFunc {
 	}
 }
 
+func (b *SystemBackend) getRaftBootstrapChallengeAnswer(serverID string) []byte {
+	// We take our local key and the given serverID and perform an HKDF
+	// invocation on it. This output is then encrypted using the root key
+	// (which protects our barrier keyring) and the remote peer can
+	// "prove" they're allowed to join by unsealing the challenge answer
+	// and providing it back to us. We assume inverting HKDF is hard (it
+	// is at least as hard as inverting an HMAC) and it is hard for an
+	// attacker to find (serverID, answer) such that:
+	//
+	//     HKDF(challengeKey, serverID) == answer
+	//
+	// The benefit of this is that we only need constant memory (a single
+	// root challenge key) and HKDF is fast and arbitrarily extensible.
+
+	kdf := hkdf.New(sha256.New, b.Core.pendingRaftPeerChallengeKey, []byte("openbao-raft-peer-challenge"), []byte(serverID))
+
+	// hkdf cannot fail on this short of output. While this value was
+	// previously 16 bytes, the root key is 32 bytes so 24 bytes provides
+	// increased safety while still retaining domain safety from the root
+	// key.
+	var answer [24]byte
+	_, err := kdf.Read(answer[:])
+	if err != nil {
+		panic(fmt.Sprintf("hkdf failed on 24 bytes of output: %v", err))
+	}
+
+	return answer[:]
+}
+
 func (b *SystemBackend) handleRaftBootstrapChallengeWrite() framework.OperationFunc {
 	return func(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
 		serverID := d.Get("server_id").(string)
 		if len(serverID) == 0 {
 			return logical.ErrorResponse("no server id provided"), logical.ErrInvalidRequest
 		}
-
-		var answer []byte
-		answerRaw, ok := b.Core.pendingRaftPeers.Load(serverID)
-		if !ok {
-			var err error
-			answer, err = uuid.GenerateRandomBytes(16)
-			if err != nil {
-				return nil, err
-			}
-			b.Core.pendingRaftPeers.Store(serverID, answer)
-		} else {
-			answer = answerRaw.([]byte)
+		if len(serverID) > 16384 {
```

Comment:
I wanted to make sure we bound the server identifier so that we didn't have endless HMAC calls (which, admittedly, are rather fast especially with SHA-256 acceleration). Previously we were upper bounded only [by `max_request_size`](https://openbao.org/docs/configuration/listener/tcp/) set at 32MB (!!!) on the length of `serverID`, which meant an attacker could cause rather large entries in the original `sync.Map`.

`localID` in `physical/raft` doesn't really have a maximum length. It is a string type value, which is used as part of a metric label and which is part of the underlying Raft protocol JSON messages. It also used [in maps](https://github.com/hashicorp/raft/blob/2b0032e390b888b0058d221989886d090b9d9609/commitment.go#L20) in Raft, so having _some_ reasonable bound seems OK.

I chose 2^14 as a reasonably large value that ~nobody should hit unless they're doing something _really_ weird with ServerID and putting a lot of... interesting... data in the field.

Happy to pull it out into a constant though.

Thoughts?

Reply:
I agree that 32 MiB is a bit much. I think it would be great if you could move the limit to a constant and document the reasoning behind it

Reply:
Perfect, pushed! Thanks @JanMa!